text/0000-hidden-impls.md

```diff
@@ -0,0 +1,878 @@
+fn main() {
+    use crate_a::Property; // ← assume that `Property::function` exists.
+
+    let x = Thing::function();
+}
+```
+
+Here we want to avoid the future possible breakage of `Property for Thing`
+eventually becoming public. Since the compiler is aware of `Property for Thing`
+and that we've brought `Property` into scope, a warning will be emitted
+in addition to the `unused import` lint, suggesting that you should use
+*uniform function call syntax (UFCS)* instead:
+
+```rust
+warning: the trait bound `Thing: Property` is not satisfied
+ --> src/main.rs:<line>:<column>
+  |
+L |     let x = Thing::function();
+  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ a hidden implementation of trait `Property`,
+  |                                which has a function `function`, exists for `Thing`.
+  |
+  | help: rewrite using UFCS:
+  |
+  |     let x = <Thing as Foo>::function();
+  |
+  = note: #[warn(hidden_fn)] on by default
+```
+
+## Hidden, but coherent
+
+It is crucial to note that `Property for Thing` acts, with respect to coherence,
+as if it were public. Consider that a public trait `Property` is defined in
+crate A and that a blanket implementation exists in crate A:
+
+```rust
+// crate A:
+
+pub trait Property<T> { ... }
+
+crate impl<T> Property<T> for T { ... } // `From` provides a blanket impl like this.
+```
+
+That means that crate B, which is a reverse dependency of crate A, may not write:
+
+```rust
+// crate B:
+
+struct Thing { ... } // or `union` / `enum`
+
+impl Property<Thing> for Thing { ... }
+```
+
+This implementation is rejected, by the compiler, as overlapping with the one
+specified in crate A. The following is one *possible* error message,
+which is a variant of `E0119`, tailored for this specific case:
+
+```rust
+error[E0119]: conflicting implementations of trait `crate_A::Property` for type `Thing`:
+ --> src/main.rs:<line>:<column>
+  |
+L | impl Property<Thing> for Thing {
+  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  |
+  = note: conflicting implementation in crate `A`:
+          - crate impl<T> crate_A::Property<T> for T { ... }
+    note: the implementation exists, but is hidden.
+```
+
+[unsound-named]: https://internals.rust-lang.org/t/looking-for-rfc-coauthors-on-named-impls/6275/2
+
+If the compiler did not reject this implementation, it would amount to the same
+problem as with named implementations, which would be decidedly unsound in Rust
+since [*"coherence is an invariant that unsafe code is allowed to assume"*][unsound-named].
+
+## Hiding things from yourself
+
+Thus far, we've only used `crate impl ..` to hide implementations.
+While we think that will be the most likely usage, we also believe
+that more fine grained control is beneficial for larger projects
+with many people involved in the development of those projects.
+Therefore, this RFC proposes that all visibility modifiers (such as
+`pub(super)`) be permitted on `impl` to hide implementations from
+different parts of the crate.
+
+## `pub` by default
```

Comment:
This is, to my mind, the biggest drawback of the proposal. I don't see a good way around it, though.

Reply:
Yup, I agree on this being a drawback :) I added a paragraph discussing it in the section on Drawbacks.